### PR TITLE
DAPI-128 DAPI-130 move delius specific config to community-api and include sentence id in referral sent request

### PR DESCRIPTION
**Pull request by mahalrmoj**
## What does this pull request do?

Moves delius specific mappings into community-api (DAPI-128), and provides sentenceId to community-api as part of the ReferralSentRequest so it can used when creating a NSI in Delius
1) Moves the responsibility of the assignment of these NSI codes (on ReferralSent) to community-api.   
    provider-code: CRS
    staff-code: CRSUATU
    team-code: CRSUAT
2) Referral type notion is replaced by service category

## What is the intent behind these changes?

A NSI Referral in Delius requires a sentenceId, building on DAPI-128 which ensures all Delius specific codes and transformations are isolated to and are the concern of community api.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIService.kt [hunk 1]
```diff
@@ -5,23 +5,18 @@ import mu.KLogging
...
 import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.events.ReferralEvent
 import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.events.ReferralEventType
 import java.time.LocalDate
 
 @Service
 class CommunityAPIService(
-  @Value("\${community-api.contact-notification-context.provider-code}") private val providerCode: String,
-  @Value("\${community-api.contact-notification-context.referral-type}") private val referralType: String,
-  @Value("\${community-api.contact-notification-context.staff-code}") private val staffCode: String,
-  @Value("\${community-api.contact-notification-context.team-code}") private val teamCode: String,
```
Comment (sldblog) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIService.kt:20]: 🙋 These are also mentioned in the configuration and helm values files -- can we remove them from those?

**Comment by sldblog** [comment]
> A NSI Referral in Delius requires a sentenceId, building on DAPI-128 which ensures all Delius specific codes and transformations are isolated to and are the concern of community API.

_not a blocker_
Could we also expand on why, please? Ultimately, what is the benefit for their users?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIService.kt (hunk 1 above)
Comment (mahalrmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIService.kt:20]: Thanks for pointing that out. Searched the code base/git for provider-code - no refs. Where are they located?

Comment (rgforsyth) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIService.kt:20]: helm_deploy/hmpps-interventions-service/templates/_envs.tpl
helm_deploy/values-dev.yaml

They are referenced such as: COMMUNITYAPI_CONTACTNOTIFICATIONCONTEXT_TEAMCODE

Comment (mahalrmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIService.kt:20]: Found it and done thanks

File: src/main/resources/db/migration/V1_33__referral_sentence_id_add.sql [hunk 2]
```diff
@@ -0,0 +1 @@
+alter table referral add column sentence_id bigint null;
```
Comment (rgforsyth) [inline src/main/resources/db/migration/V1_33__referral_sentence_id_add.sql:1]: Linter may complaint about no newline here?

**Review by rgforsyth — APPROVED** [review]
LGTM. (minor comment about newline)

File: src/main/resources/db/migration/V1_33__referral_sentence_id_add.sql (hunk 2 above)
Comment (mahalrmoj) [reply src/main/resources/db/migration/V1_33__referral_sentence_id_add.sql:1]: Thanks will check.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/CommunityAPIClient.kt [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+) {
+  companion object : KLogging()
+
+  fun makeRequest(uri: String, requestBody: Any) {
+    communityAPIWebClient.post().uri(uri)
+      .body(Mono.just(requestBody), requestBody::class.java)
+      .retrieve()
+      .bodyToMono(Unit::class.java)
+      .onErrorResume { e ->
+        logger.error("Call to community api failed for: $requestBody", e)
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/CommunityAPIClient.kt:21]: struct args would be good here

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/CommunityAPIClient.kt [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+import mu.KLogging
+import org.springframework.beans.factory.annotation.Qualifier
+import org.springframework.stereotype.Component
+import org.springframework.web.reactive.function.client.WebClient
+import reactor.core.publisher.Mono
+
+@Component
+class CommunityAPIClient(
+  @Qualifier(value = "communityApiWebClient") private val communityAPIWebClient: WebClient,
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/CommunityAPIClient.kt:11]: qualifier is not needed if the var name matches (just the capitalization is off)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/CommunityAPIClient.kt [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+) {
+  companion object : KLogging()
+
+  fun makeRequest(uri: String, requestBody: Any) {
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/CommunityAPIClient.kt:15]: maybe worth specifying this is a POST request.

**Review by tomsmyers — APPROVED** [review]
no blockers - looks good. thanks.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/CommunityAPIClient.kt (hunk 5 above)
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/CommunityAPIClient.kt:15]: may also be worth noting that this is an async call. calls may expect this to raise an exception at the call site if it fails.

Comment (mahalrmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/CommunityAPIClient.kt:15]: Brill - all for readability

**Comment by mahalrmoj** [comment]
This PR doesn't seem to have rebased ok

**Pull request closed without merging**